R/assert.R

```diff
@@ -84,3 +93,85 @@ vec_is <- function(x, ptype = NULL, size = NULL) {
 
   TRUE
 }
+
+#' Is an object a vector?
+#'
+#' @description
+#'
+#' A vector is a collection of values. In practical terms, a vector is
+#' something that can be a column of a data frame. The opposite of a
+#' vector is a scalar. Such an object cannot be [sliced][vec_slice],
+#' nor can it be a data frame column.
+#'
+#' @section Dispatch:
+#'
+#' This function is internally generic:
+#'
+#' * Atomic vectors are always vectors.
+#' * Bare lists are vectors.
+#' * Data frames are always vectors.
+#' * Non-vector base types are always scalars.
+#'
+#' Only lists can be overriden by implementing an S3 method for
+#' `vec_is_vector()`.
+#'
+#' @param x An object.
+#'
+#' @examples
+#' # Atomic vectors are always vectors:
+#' vec_is_vector(1:2)
+#' vec_is_vector(structure(1:2, class = "foo"))
+#'
+#' # Bare lists and data frames are vectors:
+#' vec_is_vector(list(1, 2))
+#' vec_is_vector(mtcars)
+#'
+#' # S3 lists need to explicitly implement `vec_is_vector()` to be
+#' # treated as lists. They are treated as scalars by default:
+#' fit <- stats::lm(disp ~ drat, mtcars)
+#' vec_is_vector(fit)
+#'
+#' # Other base types are never vectors:
+#' vec_is_vector(quote(foo(bar)))
+#' vec_is_vector(~foo)
+#' vec_is_vector(function() NULL)
+#' @export
+vec_is_vector <- function(x) {
+  switch(vec_typeof(x),
```

Comment:
I think I'd like to see this function written in C. Not because the performance is going to be different, but we're going to have a bunch of our own internal generics in the future, and this seems like a good testing ground for the basic pattern.